src/Intl/Idn/Idn.php

```diff
@@ -18,270 +36,540 @@
+        $lastLabelIndex = \count($labels) - 1;
 
-        $i = 0;
-        $length = mb_strlen($input, 'utf-8');
-        while ($h < $length) {
-            $m = $codePoints['nonBasic'][$i++];
-            $delta += ($m - $n) * ($h + 1);
-            $n = $m;
+        // Step 4. Convert and validate each label in the domain name string.
+        foreach ($labels as $i => $label) {
+            $validationOptions = $options;
 
-            foreach ($codePoints['all'] as $c) {
-                if ($c < $n || $c < 128) {
-                    ++$delta;
-                }
-                if ($c === $n) {
-                    $q = $delta;
-                    for ($k = 36;; $k += 36) {
-                        $t = self::calculateThreshold($k, $bias);
-                        if ($q < $t) {
-                            break;
-                        }
-
-                        $code = $t + (($q - $t) % (36 - $t));
-                        $output .= self::$encodeTable[$code];
-
-                        $q = ($q - $t) / (36 - $t);
-                    }
-
-                    $output .= self::$encodeTable[$q];
-                    $bias = self::adapt($delta, $h + 1, ($h === $b));
-                    $delta = 0;
-                    ++$h;
+            if (\substr($label, 0, 4) === 'xn--') {
+                try {
+                    $label = Punycode::decode(\substr($label, 4));
+                } catch (PunycodeException $e) {
+                    $info->addError(self::ERROR_PUNYCODE);
+
+                    continue;
                 }
+
+                $validationOptions['Transitional_Processing'] = false;
+                $labels[$i] = $label;
             }
 
-            ++$delta;
-            ++$n;
+            self::validateLabel($label, $info, $validationOptions, $i > 0 && $i === $lastLabelIndex);
         }
 
-        $output = 'xn--'.$output;
+        if ($info->isBidiDomain() && !$info->isValidBidiDomain()) {
+            $info->addError(self::ERROR_BIDI);
+        }
 
-        return \strlen($output) < 1 || 63 < \strlen($output) ? false : strtolower($output);
+        // Any input domain name string that does not record an error has been successfully
+        // processed according to this specification. Conversely, if an input domain_name string
+        // causes an error, then the processing of the input domain_name string fails. Determining
+        // what to do with error input is up to the caller, and not in the scope of this document.
+        return $labels;
     }
 
-    private static function listCodePoints($input)
+    /**
+     * @see https://www.unicode.org/reports/tr46/#ToASCII
+     *
+     * @param string $domainName
+     * @param int $options
+     * @param int $variant
+     * @param array $idna_info
+     *
+     * @return string|false
+     */
+    public static function toAscii($domainName, $options = \IDNA_DEFAULT, $variant = \INTL_IDNA_VARIANT_UTS46, &$idna_info = array())
```

Comment:
The public API should be kept before the private methods, to respect the Symfony coding standards.

Reply:
Thanks for letting me know. I'm sure I'm violating quite a few Symfony coding standard rules as all the code is following PSR-12. I'll have to see about running it through php-cs-fixer after I push some of the upcoming changes.